Detectors/EMCAL/simulation/src/DigitsWriteoutBuffer.cxx

```diff
@@ -20,62 +20,104 @@
+    auto towerEntry = buffEntry.mDigitMap->find(towerID);
+    if (towerEntry == buffEntry.mDigitMap->end()) {
+      towerEntry = buffEntry.mDigitMap->insert(std::pair<int, std::list<o2::emcal::LabeledDigit>>(towerID, std::list<o2::emcal::LabeledDigit>())).first;
+    }
+
+    buffEntry.mTimestamp = dig.getTimeStamp();
+
+    // if the digit time is equal to the mLiveTime, the time bin is set to be the end time window
+    // if the digit time is between mLiveTime and busy time the time bin is set to not record mode
+    // if the digit time is > (mLiveTime + mBusyTime - mPreTriggerTime), the time bin will be set to record mode
+    if (dig.getTimeStamp() == mLiveTime) {
+      buffEntry.mRecordMode = false;
+      buffEntry.mEndWindow = true;
+    } else if (dig.getTimeStamp() >= mLiveTime && dig.getTimeStamp() < (mLiveTime + mBusyTime - mPreTriggerTime)) {
+      buffEntry.mRecordMode = false;
+      buffEntry.mEndWindow = false;
+    } else if (dig.getTimeStamp() < mLiveTime || dig.getTimeStamp() >= (mLiveTime + mBusyTime - mPreTriggerTime)) {
+      buffEntry.mRecordMode = true;
+      buffEntry.mEndWindow = false;
+    }
+
+    towerEntry->second.push_back(dig);
   }
+}
 
-  towerEntry->second.push_back(dig);
+// When the current time is forwarded (every 100 ns) the entry 0 of the future dequeue
+// is removed and pushed at the end of the past dequeue. At the same time a new entry in
+// the future dequeue is pushed at the end, and - in case the past dequeue reached 15 entries
+// the first entry of the past dequeue is removed.
+void DigitsWriteoutBuffer::forwardMarker(double eventTime)
+{
+
+  int sampleDifference = int(eventTime / 100) - int(mLastEventTime / 100);
+
+  for (int idel = 0; idel < sampleDifference; idel++) {
+    if (mTimedDigitsFuture.front().mEndWindow || !mTimedDigitsFuture.front().mRecordMode) {
+      break;
+    }
+
+    mTimedDigitsPast.push_back(mTimedDigitsFuture.front());
+    mTimedDigitsFuture.pop_front();
+    mTimedDigitsFuture.push_back(DigitTimebin());
+
+    if (mTimedDigitsPast.size() > mBufferSize) {
+      mTimedDigitsPast.pop_front();
+    }
+  }
 
   mLastEventTime = eventTime;
 }
 
-std::list<std::unordered_map<int, std::list<LabeledDigit>>> DigitsWriteoutBuffer::getLastNSamples(int nsamples)
+// TODO Pre-trigger collisions still can't be added with the others
+// Reset their time
+std::list<std::unordered_map<int, std::list<LabeledDigit>>> DigitsWriteoutBuffer::getLastSamples()
 {
-  int startingPosition = int(mLastEventTime / mTimeBinWidth) - nsamples;
-  if (startingPosition < 0) {
-    startingPosition = 0;
-  }
 
   std::list<std::unordered_map<int, std::list<LabeledDigit>>> output;
 
-  for (int ientry = startingPosition; ientry < startingPosition + nsamples; ientry++) {
-    output.push_back(mTimedDigits[ientry]);
+  for (auto digitTimebin : mTimedDigitsPast) {
+    output.push_back(*digitTimebin.mDigitMap);
   }
```

Comment:
Consider a gsl::span (non-owning as return type in order to avoid the copy process:
``` c++
const gsl::span<DigitsWriteoutBuffer::DigitTimeBin> DigitsWriteoutBuffer::getLastSamples() const 
{
   return const gsl::<DigitsWriteoutBuffer::DigitTimeBin>(mTimedDigitsPast);
}

Reply:
Done